### ISSUE TITLE
Introducing @dir

### ISSUE THREAD

**Issue by gkellogg** [w3c/json-ld-syntax#11]
In some situations it is important/necessary to include the base direction of a text, alongside its language; see the [“Requirements for Language and Direction Metadata in Data Formats”](https://w3c.github.io/string-meta/) for further details. In practice, in a vanilla JSON, it would require something like:

  ```
  "title": [ { "value": "Moby Dick", "lang": "en" },
	         { "value": "موبي ديك", "lang": "ar"  "dir": "rtl"}
	       ]  
  ```

  (the example comes from that document).

  At this moment, I believe the only way you can reasonably express that in JSON-LD is via cheating a bit:

  ```
  "title": [ { "@value": "Moby Dick", "@language": "en" },
	         { "@value": "موبي ديك",  "@language": "ar"  "dir": "rtl"}
	       ]  
  ```

  and making sure that the `dir` term is _not_ defined in the relevant `@context` so that, when generating the RDF output, that term is simply ignored. But that also means that there is no round-tripping, that term will disappear after expansion.

  The difficulty lies in the RDF layer, in fact; RDF does not have any means (alas!) to express text direction. On the other hand, this missing feature is a general I18N problem whenever JSON-LD is used (there were issues when developing the Web Annotation Model, these issues are popping up in the Web Publication work, etc.).

  Here is what I would propose as a non-complete solution

  1. Let us introduce a `@dir` term, alongside `@language`. This means this term can be used in place of `dir` above, ie, it is a bona-fide part of a string representation, and would therefore be kept in the compaction/expansion steps, can also be used for framing.
  2. In JSON-LD 1.1, `@dir` is _ignored_ when transforming into RDF. I.e., only the language tag would be used.
  3. We _may_ initiate some work in the RDF community to solve this issue. There may be several ways, each of them require the RDF community to chime in
  	3.1. Define a mechanism of "parametrized" standard datatypes that represent a (language,direction) pair. One would then get something like`[] ex:title "موبي ديك"^^rdf:internationalText(ar,rtl) ;`
  	3.2. Go for a "generalized" RDF where strings can also appear as subjects (that has been a matter of dispute for a long time...). That would give the possibility to add such attribute to texts like directions
  	3.3. Some other mechanisms that I cannot think about
  4. In a future JSON-LD 1.* the `@dir` value can be properly mapped onto an RDF representing the right choices (if such choices are worked out)

Original issue [Introducing @dir ? #583](https://github.com/json-ld/json-ld.org/issues/583)

**Comment by iherman** [comment]
As the one who raised this: I am in favour of closing this.

**Comment by azaroth42** [comment]
Closing, wontfix, as new features should be compatible with the RDF data model.
WG Resolution: https://www.w3.org/2018/json-ld-wg/Meetings/Minutes/2018/2018-08-03-json-ld#section2-2

**Comment by azaroth42** [comment]
To be discussed at TPAC 2019, given lack of interest in a separate WG to fix text direction in RDF generally

**Comment by iherman** [comment]
This issue was discussed in [a meeting](https://www.w3.org/2018/json-ld-wg/Meetings/Minutes/2019/2019-09-19-json-ld#section5).

- `RESOLVED:  (1/) We add <code>@direction</code> as a keyword to the JSON-LD Syntax to assert the base text direction of a literal string*`
- `RESOLVED:  (2/) <code>@direction</code> is legal as part of a value object in the json-ld internal model but may not be used if <code>@type</code> is present*`
- `RESOLVED:  (3/) <code>@direction</code> is legal in an expanded term definition if you could have used <code>@language</code> in that definition and applies to all values of that term in the same way as <code>@language*</code>`
- `RESOLVED:  (4/) The value space of <code>@direction</code> is one of “ltr” and “rtl” (“auto” would have been the same as not asserting the keyword)*`
- `RESOLVED:  (5/) We expand the possible values of language maps to be either a string (as per today) or a value object with direction and without language or type*`
- `RESOLVED:  (6/) <code>@direction</code> can be present in a context node, and sets in the active context that the default direction for all string literals*`
- `RESOLVED:  (7/) WG agrees that <code>@container:</code> <code>@direction</code> is not a useful feature for JSON-LD 1.1 (unlike <code>@container:</code> <code>@language)*</code>`
<details><summary><i class="differentiate">View the transcript</i></summary>
Rob Sanderson:<br/>
{<br/>
<b>"<code>@context":</b></code> {<br/>
<b>"<code>@vocab":</b></code> "<a href="https://example.com/",">https://example.com/",</a><br/>
<b>"schema":</b> "<a href="https://schema.org/",">https://schema.org/",</a><br/>
<b>"name":</b> "schema:name"<br/>
},<br/>
<b>"name":</b> {<br/>
<b>"<code>@value":</b></code> "fish",<br/>
<b>"<code>@language":</b></code> "en",<br/>
<b>"<code>@direction":</b></code> "ltr"<br/>
}<br/>
}<br/>
Pierre-Antoine Champin: +1<br/>
Pierre-Antoine Champin: { “<code>@context”:</code> [“<a href="http://schema.org/”,">http://schema.org/”,</a> { “<code>@direction”:</code> null }], “tag:p”: { “<code>@value”:</code> “foo”, “<code>@language”:</code> “en”, “<code>@direction”:</code> “ltr” } }<br/>
Manu Sporny: Discussion around use of <code>@direction</code> and how it affects JSON-LD 1.0 processors.<br/>
Manu Sporny: Seems that there is an issue wrt. the use of <code>@vocab</code> and <code>@direction…</code> processors are choking.<br/>
Manu Sporny: A bug exists in 1.0 and 1.1 that doesn’t flag the use of <code>@direction</code> and <code>@vocab.</code><br/>
Rob Sanderson: Issue created: <a href="https://github.com/w3c/json-ld-syntax/issues/259">https://github.com/w3c/json-ld-syntax/issues/259</a><br/>
Manu Sporny: Current 1.0 processors are going to ignore use of <code>@direction</code> in various places.<br/>
Pierre-Antoine Champin: {"<code>@version":</code> null} :-D<br/>
Rob Sanderson: Moving on to whether or terms can start with @ symbol.<br/>
… Currently, the spec is not clear wrt. SHOULD NOT vs. MUST NOT.<br/>
Benjamin Young: This is why HTML is not versioned.<br/>
… I’m less concerned about two documents processing two documents differently… we don’t want to touch old data, but we don’t want to prevent accomodation of the future at the same time.<br/>
Gregg Kellogg: If HTML pages had something that wasn’t interpreted, but then it would… like span, or blink, or marquee … then there could be issues.<br/>
Benjamin Young: When I use a newer processor, I’m aware that it’s going to look different… I don’t touch old data.<br/>
… I’d suggest you use a switching mechanism in processor… nuclear option is to use 1.1.<br/>
… If I point towards the past… I provide HTML5 document to HTML 3 processor, you don’t lose data<br/>
… If you give <code>@version</code> to 1.0 processor, it explodes.<br/>
Manu Sporny: Yes, but the only consumers are schema.org and they’ll just do 1.1.<br/>
Benjamin Young: yes, the evergreen browser argument<br/>
Pierre-Antoine Champin: Changing 1.0 parsers, if they don’t fail on 1.1 parsers, they’ll break on other things<br/>
Benjamin Young: See Feature Detection issue<br/>
Gregg Kellogg: We have to do something intelligent when we hit <code>@container:</code> subway – 1.0 will throw a syntax error, it should throw errors when it sees an issue. My feeling is that we should remove the explicit version announcement, and do feature detection when things are discovered and throw errors when they don’t understand.<br/>
Ivan Herman: Another reason why I’m in favor of incremental stuff, in a weaker way than what you say, is because we have to look ahead, we’ll have 1.1 now, but there’ll be a different discussion on how to proceed when this WG stops… listening to presentation from yesterday, next charter whatever 2.0 - essentially, admin and names put aside, we’d be in a position to put one feature that we want to add to the language like direction.<br/>
… If in a year from now, we add <code>@direction,</code> what do we do – version 1.1.1 or version 1.2?<br/>
… We get into versioning/naming issue… it’s difficult to maintain… sometimes they release 5.3, they do incremental changes, then implementations catch up… we might have the same sort of future ahead of us, we should be careful to not make things too difficult… 1.1 as a standard adds new features, not necessarily a thing called a 1.1 processor, we hope we have processors that implement everything, that’s the goal.<br/>
Manu Sporny: so. the original <code>@version</code> proposal was a list of features<br/>
… so if you were to process this document, you must support those features<br/>
… if a processor can’t support those, then it just blows up<br/>
… and 1.1 was just an alias for a bundle of features<br/>
… the issue we saw was that there might be a feature…well…<br/>
… one, announcing them doesn’t mean you’ll be using them<br/>
… or it might just get forgotten that you’ve left the “opt-in” around when you ended up not using it after all<br/>
… so, the feature detection works best when it’s also BC friendly<br/>
… and that turns out to be rather hard<br/>
… we’d have to go through each feature and determine how detectable they are<br/>
… but the reason against that is that it’s a huge amount of spec text<br/>
… and we weren’t sure that every new feature weren’t able to figure those out<br/>
… and they might determined it differently<br/>
… which is why we were ultimately in favor of 1.1<br/>
Ivan Herman: so does that mean if we go into continuous development<br/>
Manu Sporny: …yeah. I’m against that also<br/>
… I was opposed to this 1.1 work earlier…the 4 year gap was helpful for the stability of the data and community<br/>
Ivan Herman: k…but if we keep shipping, then we’d just keep shipping these version number things<br/>
Manu Sporny: yeah…but that gets complicated again<br/>
… and if you look at HTML5, feature detection is insane<br/>
… the earlier issues were about feature announcement not detection<br/>
Gregg Kellogg: the point of 1.1 was not so we’d have a 1.2<br/>
… it was so we’d be locking things down<br/>
… and that’s principally done in context processing<br/>
… and we did not do that properly in 1.0<br/>
… so, virtually, I’d say 90% of features are discovered in <code>@context</code> processing<br/>
… The original version was 1.1, was not to do 1.2, it was so we could detect things that should not be there in 1.0… if 1.2 added <code>@direction,</code> a 1.1 processor would fail if it saw it because it was not in the whitelist. 1.2 is not necessary to say <code>@version</code> 1.2 to do the right thing… virtually, 90% of feature issues are discovered in context processing, the exception is when processing expanded documents where there are some features used.<br/>
… the others show up in expansion<br/>
… We do look for those in 1.1 mode and 1.0 mode… frame processing mode…<br/>
… I think where we are detecting these things in expansion processing, maybe more needed for forward compatability, everything else is pretty much be in the <code>@context</code> processing phase.<br/>
Ivan Herman: If we had it today, <code>@direction</code> wouldn’t be detected in <code>@context.</code><br/>
Gregg Kellogg: If we had it in the context it would be.<br/>
Ivan Herman: If I just use it as part of the value, it’s a 1.1 feature that I don’t see in the <code>@context.</code><br/>
Gregg Kellogg: Anything that looks like a keyword and the processor doesn’t understand it, it should abort.<br/>
… I disagree w/ manu, I think we can do feature detection.<br/>
Manu Sporny: I don’t think I said we’d use 1.1 when doing JSON-LD 5.0<br/>
Ivan Herman: isn’t this an issue for stream processor?<br/>
Gregg Kellogg: Well, you could throw an error if you hit something…<br/>
… I think we’d be in a better position if we don’t use <code>@version.</code><br/>
<b>rob:</b> We need to break for lunch soon.<br/>
Sanders Tolk: In some cases, it’s important to break down with newer features used… it’s a conscious choice to move from a 1.0 to a 1.1 interpreter…<br/>
… There is something to be said to an interpreter breaking on a feature it doesn’t understand, vs. a feature thing… parser, or implementation…<br/>
… Depending on level of confidence, you use one or the other… this is something that people should be aware of anyway.<br/>
Gregg Kellogg: In some cases, in order to get behavior, you should be in validation mode.<br/>
Benjamin Young: I want as much of data out of top of page, don’t care about reviews, give me what you can give me.<br/>
Sanders Tolk: To be able to post a snippet in a web page, does this have any unknown features for 1.1… so people can know, am I going outside of that…<br/>
Gregg Kellogg: It is using features that are defined in 1.1… if your parser dies, sometimes it helps if you have it all gathered.<br/>
6. Text directions, Part II<br/>
Rob Sanderson: The issue that we want to talk about is notion of internationalization and feature of text direction of strings, which is not possible in underlying RDF data model. Proposal that came out of discussion was that it would be better for the infrastructure overall if JSON-LD did text direction. Absence of it already existing at RDF concepts layer… it would be better back port it into RDF from JSON-LD.<br/>
Addison Phillips: We move the upper stories of the Eiffel tower first :)<br/>
… We are concerned that there is a window opened right now that so many specs are dependent on JSON-LD, we would like to make sure common serialization form for handling this, if we don’t have something, we might have non-interop solutions for this.<br/>
… We are interested in seeing if we can do something at this level, see if we can move this down to RDF level.<br/>
Rob Sanderson: One question that would be good to answer, why isn’t language sufficient to do this?<br/>
… Why can’t we just say English is ltr, arabic is rtl, and be done…<br/>
Addison Phillips: You can infer base direction, but not the same as base direction information, like when you don’t have accurate language metadata, or applying base direction isn’t applicable.<br/>
Richard Ishida: See Article on Base direction and language tag<br/>
Addison Phillips: Rather than doing coercion, we should have metadata slot rather than interfering… that’s the basis of it. Key aspect is that it has to do w/ mixed directional text… text that flows in one direction isn’t a problem, text that starts or ends with an opposite direction, bring multiple strings together, you have spillover effects that are desirable. That’s a challenge that we see as fairly fundamental.<br/>
Charles Neville: There is another problem in RDF in particular, which is that RDF doesn’t know anything about the language… lang=ar, you know it’s rtl… the only modern language that you can’t detect is azeri.<br/>
Richard Ishida: No, not true by a long shot.<br/>
Charles Neville: RDF doesn’t understand ar-kw ar-?? is all arabic<br/>
Richard Ishida: And new subtags are added all the time…<br/>
Charles Neville: You’d have to parse RDF language tags to get this partial solution<br/>
Richard Ishida: There was a point where we thought it would work… the URL that I just posed in IRC is the result of that.<br/>
… The conclusion we came to was that language wouldn’t do the job… wrote two articles on Kashmiri - the language tagging that I used in both documents was ks, didn’t put script info because I didn’t need to… there is a lot of human stuff going on there. You don’t know if it’s rtl or ltr. Even I didn’t feel like I needed to put subscript and relying on language.<br/>
Ivan Herman: Coming back to what Addison said, whatever we do, there are cases that we cannot handle directly, and we shouldn’t even go there. If we have text that internally mixes language, the only way we could handle that is to fall back on HTML encoding of content.<br/>
… metadata only on the one string.<br/>
… When you have mixed text, we can’t do that.<br/>
Addison Phillips: We just want base direction of whole string… there are bidi things that can happen on whole string. Supplied in form of HTML… set the base here… we’re not talking about internal changes<br/>
Dan Brickley: What RDF understands of the language structure… a lot of stuff you just pass through RDF… if you want to be able to write SPARQL queries, you’d have to have advanced knowledge of those language fields, you push the solution into the application.<br/>
Addison Phillips: Language tag mapping, don’t know what the subtags mean… SPARQL just does string matching, if you say ‘bs’ which stands for “bosnian”, you need a “bs range” to match.<br/>
… You still don’t need to know what the tags and subtags mean… the hyphens tell you what the boundaries mean… it’s deliberate, keep away from needing to read registry, use complex logic.<br/>
… If you are doing complex matching, you need application logic.<br/>
Dan Brickley: We don’t need to address fancy SPARQL query.<br/>
Sandera Stolk: I was going to talk about lang matches, doesn’t fix solution for direction… you’re probably aware of, if a solution is found for direction in JSON-LD, there may be multiple other types of language interpretation attributes to help with something that goes beyond direction… if there is a solution found in JSON-LD, it should also be a solution that should allow for other attributes as time goes by… it can be placed into the same construct.<br/>
Charles Neville: If we follow this plan, what breaks?<br/>
Ivan Herman: What’s the plan? Add syntax for <code>@direction</code> and add a new property<br/>
Charles Neville: See the proposed json-ld syntax<br/>
Manu Sporny: Group looks at directions in rdf-dir-literal document.<br/>
Ivan Herman: Note that the document we are referring to (link above) describes a range of possible solutions. We suggested implementing the json-ld syntax in json-ld as a starting point, because it doesn’t yet have a representation in RDF semantics<br/>
… We ended up saying this, if I put RDF hat on, there are downsides in practice, there are duplicated ways to represent literals in RDF… literal w/ language that RDF has today, in parallel we have another approach, might break some deployed OWL mechanisms, object properties, data properties, not a good solution, but possible solution…<br/>
Gregg Kellogg: We also discussed that this was a transitional solution.<br/>
Ivan Herman: Hope is in time, there is a more general mechanism in RDF, don’t know if this will happen. Current proposed approach, maps nicely to JSON-LD, covers what sander wants.<br/>
… What we discussed today, yes, this is doable, but what happens w/ existing JSON-LD data out there, and what are the effects… and answer is not that simple…<br/>
… Adding the feature is easy, what happens to the process is not a big deal….<br/>
Manu Sporny: I really pushed to fix this, for verifiable credentials to be able to move on…<br/>
… we rushed in to getting good ideas on the table. I am going to check whether this is really the right course.<br/>
… seems like language tag does 95% of what we want, a lot of times.<br/>
… so it is true that people extract some text direction out of the language tag<br/>
Addison Phillips: mostly right now it isn’t used…<br/>
… most people without a base direction appy the unicode algorithm.<br/>
Charles Neville: anecdata suggests that in practice, people are really implementing stuff withthe language tag and accept that as generally good enough for customers.<br/>
Manu Sporny: There is a reality on the ground that people are using the lang tag. All of these specs getting changed might not be as good as working out how to do this through teh lang tag.<br/>
… we could deal with it in the lang tag and not have to change the RDF layer<br/>
Benjamin Young: the mentioned list of specs <a href="https://w3c.github.io/rdf-dir-literal/draft-charter.html#normative">https://w3c.github.io/rdf-dir-literal/draft-charter.html#normative</a><br/>
Addison Phillips: One thing is that by design implemnentations don’t need to know about the content of the language tag - and most of them don’t.<br/>
… there is a massive number of subtags.<br/>
Addison Phillips: One of the things I mentioned earlier, by design, implementations don’t need to know about subtags/language tags… we can maybe smuggle serialization schemes… the direction metadata doesn’t belong in the language tag, it’s semiorthogonal not totally unrelated, separate piece of data, I’ll admit that people have struggled… we failed to provide a mechanism to do this, we should have had this conversation before the paint was dry.<br/>
… we have looked, in <a href="https://w3c.github.io/rdf-dir-literal">https://w3c.github.io/rdf-dir-literal</a> at ways to “smuggle” the information metadata in via the lanugage tag. It is semi-orthogonal, being a separate bit of data, that we have to be able to serialise and deserialise. There are a lot of places we suffer because we haven’t yet managed to solve the problem, since we didn’t do it from the start.<br/>
Richard Ishida: r12a’s slides start here: <a href="https://www.w3.org/International/talks/1810-paris/#metadata">https://www.w3.org/International/talks/1810-paris/#metadata</a><br/>
Addison Phillips: People who have data catalogs who have base direction metadata, when people have this affordance, it helps to be able to do this… does it work a lot of the time, yes, but corner cases are really not that unusual…<br/>
… people with data catalogues who have the metadata have an easy time if they can set the direction directly, Carrying around lists of subtag matching works a lot of the time but is pretty ugly, and the corner cases are not that unusual.<br/>
… we launched a site in an RTL language, and hit a pile of bugs. I don’t always have convenient mechanisms to manage this, so I am interested in a way to provide direction metadata in JSON-LD<br/>
… Where I’m headed to here, I am not a purist, if we had an impure serialization scheme that had greater compat, that wouldn’t be bad…<br/>
… Talking w/ RDF community, we’re creating an interesting hack, doesn’t speak to extensibility… don’t know if it’s a hack… I’d be open to a serialization, looks hacky relative to other alternatives… maybe we can be compatible w/ existing implementation things.<br/>
Dan Brickley: I liked the idea of doing something in JSON-LD… if we could backport for 1.0, we could get something happening in 1.0, Google might get involved.<br/>
… We’ve encouraged people do to good solid JSON-LD 1.0 files, there are millions and millions of this….<br/>
… The smallest little thing we can do is important, if we can do some surface hackery, we can parse this stuff…<br/>
Gregg Kellogg: Yes, some transitional mechanism.<br/>
Ivan Herman: We are having a discussion about versioning discussion…<br/>
Gregg Kellogg: parameterize RDF algorithms stuff…<br/>
Dan Brickley: Parser being invoked w/ arguments, people run RDFa parser in certain mode.<br/>
Rob Sanderson: There are two separate questions… the first one, do we feel comfortable adding <code>@direction</code> to JSON-LD for <code>@value</code> objects regardless of how it transforms to RDF.<br/>
… There are things that don’t round trip in JSON-LD, such as indexes, there is precendent for non-roundtrippable stuff in JSON-LD… if cost to RDF layer is high, we can defer the serialization through into RDF… the proposal is 1) do we want to do this, to 2) how do we want to do this.<br/>
Charles Neville: I think we want to do this, happy to give up round trippability… what’s going to break?<br/>
… If that’s a fast path, this is a good solution.<br/>
… Forcing things that don’t parse language to parse language, that’s a big ecosystem thing.<br/>
Manu Sporny: I think the vast majority of appications don’t pay attention to the language tags, or direction.<br/>
Ivan Herman: In browsers, they don’t establish direction with language they do it with direction metadata.<br/>
Richard Ishida: <a href="https://github.com/w3c/string-meta/wiki/Base-direction-and-language-tags">https://github.com/w3c/string-meta/wiki/Base-direction-and-language-tags</a><br/>
Richard Ishida: The normal way of detecting language is to look at first strong character heuristics, for most things in strings, that’s the way to do it… far more efficient to look at first character… that works for 95% of time time, for the other 5%, that’s the solution we’re looking for, not having language, not having direction, and when direction on character is wrong…<br/>
Manu Sporny: I just wanted to make sure we are making an educated decision.<br/>
Addison Phillips: I think the right thing to have done is split the two – language and dir must be separated.<br/>
Richard Ishida: Yes, they’re not doing the same thing<br/>
<b>sander:</b> Pinpoint something wrt RDF - our company uses multiple language tags to discuss the same term, we find it’s difficult to find proper name for same term.<br/>
… We would be more reluctant if we adopt this that this is a different language encoding to …<br/>
Ivan Herman: If you look at this thing in RDF version, if there is no <code>@direction,</code> it maps against a literal and language tag… it’s some bnode with properties, it’s very different.<br/>
<b>sander:</b> I wanted to point out that if capturing such information would require another pattern, then such adoption for orgs to adopt direction, if that causes structural thing, that would hinder even looking for these things.<br/>
… smuggling direction in language might help adoption because it’s easier to do… let’s not make this have additional barriers.<br/>
Gregg Kellogg: To clarify, are we done at 3?<br/>
Rob Sanderson: This is something we want to do…<br/>
Gregg Kellogg: We get together we have an hour, we do the same discussion, we should put a stake in the ground. Something is better than nothing.<br/>
Manu Sporny: I’m committed and convinced to do <code>@direction</code><br/>
Rob Sanderson: Is it possible for you to be back after break, danbri?<br/>
Dan Brickley: What about the future? When you have arabic, french, in Japanese?<br/>
… Are we defining something where future direction happens…?<br/>
Charles Neville: What about text orientation and stuff… LTR and RTL…<br/>
Addison Phillips: There is a high degree of confidence on number of text directions…<br/>
Rob Sanderson: on the Agenda is continue until exhaustion<br/>
… in favor of progress<br/>
… I think that is contentious to add the surface syntax<br/>
… so reopen that original issue<br/>
… and then look at RDF round tripping after that’s resolved<br/>
… the worst scenario is to go to CR with no solution<br/>
… it is better to go CR with <code>@direction</code> and without a mapping into RDF, than to not put in <code>@direction</code><br/>
Gregg Kellogg: curious about danbri’s thoughts<br/>
Dan Brickley: that’s up to Google (Bing, etc), but we can encourage users of Schema.org to publish it<br/>
Rob Sanderson: (better to also have consensus about the RDF transitional reflection, but that seems more complicated)<br/>
Dan Brickley: but can’t promise SEO bots will consume it<br/>
Ivan Herman: if I put in a <code>@language</code> into my web site data, what happens<br/>
Dan Brickley: we certainly have the notion of languages of pages, etc.<br/>
… but the more narrow <code>@language</code> usage would be up to the various products/consumers of that data<br/>
… and I can’t speak for what they do<br/>
… I think it’s likely that we’re not doing a great job with <code>@language–let</code> alone <code>@direction,</code> etc.<br/>
… so I don’t think our tooling will catch fire, etc.<br/>
… it would be nice if there were an appendix or something that said even if you’re publishing in 1.0, that you can still put <code>@direction</code> in and hope for the best<br/>
Gregg Kellogg: there have been some calls in Schema.org to add some properties to say<br/>
… things like breadcrumbs might make use of the <code>@list</code> capability<br/>
… or strings that used a map structure<br/>
… where there would be a <code>@context</code> to understand those<br/>
… whereas at Google they’re just checking directly<br/>
… might you consider a similar pattern for <code>@direction</code><br/>
Dan Brickley: I think you’re designing hacks for a system neither of us know much about :)<br/>
Rob Sanderson: let’s project some examples<br/>
Dan Brickley: I was thinking whatever Voodoo you work up–so in 1.1 it’s these little @ things…<br/>
Gregg Kellogg: so, yes, we have <code>@language</code> and soon, we think, <code>@direction</code><br/>
… I can set an <code>@language</code> for the whole document or a certain object<br/>
… and when it gets expanded, you have that language data as well as the value<br/>
Dan Brickley: so each publishing web site would make use of its own context?<br/>
Benjamin Young: let’s make sure we’re all talking using the same terms<br/>
… I don’t we’re all sure what expanded form is, etc.<br/>
Rob Sanderson: [projects some examples]<br/>
Dan Brickley: for examples of google’s use of json-ld (and schema.org), see Search Gallery at <a href="https://developers.google.com/search/docs/guides/search-gallery">https://developers.google.com/search/docs/guides/search-gallery</a><br/>
Manu Sporny: so I feel we’re overusing an unbuilt feature<br/>
… how often to we expect people to use the <code>@direction</code> feature actually?<br/>
… most authors won’t want to grow to great pains to express it<br/>
Addison Phillips: no, but most machines written to care will<br/>
… so people with direction metadata will want to consistently use it all over the place<br/>
Dan Brickley: Can’t we just use the dir attribute on a script tag?<br/>
… You can pick it up from the HTML dir element…<br/>
Benjamin Young: No, we’re not allowed to do that.<br/>
… You can send it into processing algorithms for JSON-LD…<br/>
Ivan Herman: We looked at this, HTML5 spec says to not use lang for script tags.<br/>
Dan Brickley: <a href="https://html.spec.whatwg.org/multipage/scripting.html#data-block">https://html.spec.whatwg.org/multipage/scripting.html#data-block</a><br/>
… is there a way to do this at the top of the file?<br/>
Ivan Herman: no<br/>
Gregg Kellogg: you can at the top of the file for <code>@language</code><br/>
… and presumably we would for <code>@direction</code><br/>
Ivan Herman: so, yes, it could be there, and on each value<br/>
… it is an open question whether the container mapping thing would also be used for language direction<br/>
… basically, yesterday’s proposal was to mimic <code>@language,</code> but for the <code>@direction</code> data<br/>
Benjamin Young: clearer? or worse?<br/>
Dan Brickley: but what if context’s aren’t used?<br/>
Dan Brickley: eg <a href="https://gist.github.com/danbri/9627452ec7dcbe63af9cb5a631acfdf8#file-gistfile1-txt">https://gist.github.com/danbri/9627452ec7dcbe63af9cb5a631acfdf8#file-gistfile1-txt</a><br/>
Manu Sporny: so you’d just use <code>@language</code> and <code>@value</code> directly as an object for the value of the property–i.e. a value object<br/>
Dan Brickley: so, here’s an example we’ve been playing with at Google<br/>
… we don’t want to do this on every field–as you can see it’d be painful and verbose<br/>
Gregg Kellogg: yeah, we would let you set for the whole document–as with <code>@language</code><br/>
Manu Sporny: I feel like we’re handing people a foot gun<br/>
… everything would become ltr or rtl–even if it’s not supposed to<br/>
Gregg Kellogg: well, it depends on how they visualize the output to know if they’ve done it wrong<br/>
Manu Sporny: people don’t visualize it<br/>
Benjamin Young: they just put the JSON into text areas and call it done<br/>
Manu Sporny: yeah…most of them don’t even care about <code>@language</code><br/>
Dan Brickley: so do it only down at the property level<br/>
Addison Phillips: I can see doing it there<br/>
Dan Brickley: would it make sense to have it in the schema languages?<br/>
… like certainly fields are “directable”?<br/>
… so gsin’s or isbn’s are not directable or language-able<br/>
Addison Phillips: isbn’s are a great example<br/>
… they have to be a string, includes letters, hyphens, etc.<br/>
… but it’s not in a language<br/>
… it’s a string literal<br/>
… and the world is few of these<br/>
… but it’s also full of language-bearing strings which have both language and direction<br/>
… in some cases that can be inferred, but in some cases we can’t<br/>
… so we do have direction fields in our databases to track that<br/>
Manu Sporny: so, one option is to allow <code>@direction</code> at the top and have it cascade to the entire document<br/>
… the other extreme/option is to put it on only the values<br/>
Ivan Herman: so it is up to Schema.org to decide if they want to promote <code>@direction</code> at the top<br/>
… we are here talking about the standard<br/>
… and I don’t see why we would disallow this<br/>
Addison Phillips: the reason you do this with <code>@language</code> is because you don’t want to repeat it<br/>
Ivan Herman: the same applies to <code>@direction</code><br/>
Gregg Kellogg: we have various things to apply <code>@language</code> across the whole document<br/>
… I can say that ISBN is no language<br/>
… by setting <code>@language:</code> null on just that value<br/>
… for symmetries sake it would be odd for us to do this only for <code>@language,</code> but not <code>@direction</code><br/>
… even though we might not have obvious use cases, there is an orthogonality argument to make to do it for the whole document<br/>
Ivan Herman: I agree<br/>
Addison Phillips: I agree<br/>
Gregg Kellogg: something I can see us doing is that if there’s a default <code>@direction,</code> then it could be limited to only things that bear a <code>@language</code><br/>
Ivan Herman: no, ISBN is a good example<br/>
… they always go “rtl”<br/>
Addison Phillips: MAC addresses are a great example<br/>
… they draw backwards if they flow the wrong way<br/>
Gregg Kellogg: so, we should do this in all the same places we allow <code>@language</code><br/>
<b>group:</b> [general positive responses]<br/>
Gregg Kellogg: I presume it doesn’t make since to have <code>@direction</code> maps?<br/>
Addison Phillips: <code>@language</code> maps are very useful<br/>
… the question then is how to do both if you use <code>@direction</code><br/>
Gregg Kellogg: we could do nested objects inside nested objects<br/>
<b>group:</b> [growns]<br/>
<b>tung:</b> so we define only the language, and guess the direction?<br/>
Gregg Kellogg: you could leave the direction off<br/>
… you could provide both<br/>
… or the string could just use the default direction<br/>
<b>tung:</b> if the user provides a direction, but if the user does not state a direction it falls back to language?<br/>
Addison Phillips: the language map is the space in which you’d define a language<br/>
… but on each value<br/>
Duy-Tung Luu: what about vertical text?<br/>
Addison Phillips: vertical text is a presentational concern<br/>
Manu Sporny: I think we keep breaking our own principles here<br/>
… now we’re saying we’re using language for text direction when it’s not available?<br/>
Addison Phillips: they’re not completely independent<br/>
Manu Sporny: yeah…so the fact that this was thoroughly confusing to a bunch of developers…<br/>
… the stuff presented earlier about not guessing direction from language…but now we are<br/>
<b>uests:</b> David_Clarke<br/>
David Clarke: Yiddish can be in Romain characters or Hebrew characters<br/>
… so declaring the direction is still important<br/>
Manu Sporny: page-wide dir on HTML was maybe a mistake<br/>
Addison Phillips: no, there’s a reason to have it on the outside of the container for a bunch of reasons<br/>
… there is some linkage between language and direction<br/>
… and what we want is to give people the things they need to reach beyond what language provides<br/>
… if you ever watch folks attempting to develop or write this stuff, it’s really really hard<br/>
… we’re trying to provide enough affordance to take a string out of a page and have it present properly<br/>
… without having to introspect, etc.<br/>
… we’d like for folks to be able to express their data in all the various standards without having to rewrite all the things<br/>
… that’s why having direction is important here<br/>
… because it handles all the other bits beyond the text–bullet points, etc.<br/>
… works in html, etc.<br/>
*Proposed resolution: (1/) We add <code>@direction</code> as a keyword to the JSON-LD Syntax to assert the base text direction of a literal string** (Rob Sanderson)<br/>
Manu Sporny: +1<br/>
Gregg Kellogg: +1<br/>
Rob Sanderson: +1<br/>
Benjamin Young: it’s really similar to RDFa using a document level dir for all the things<br/>
… and being overridden elsewhere<br/>
Benjamin Young: +1<br/>
Simon Steyskal: +1<br/>
Ivan Herman: +1<br/>
Pierre-Antoine Champin: +1<br/>
Dan Brickley: +1<br/>
Resolution #3: (1/) We add <code>@direction</code> as a keyword to the JSON-LD Syntax to assert the base text direction of a literal string*<br/>
*Proposed resolution: (2/) <code>@direction</code> is legal as part of a value object in the json-ld internal model but may not be used if <code>@type</code> is present** (Rob Sanderson)<br/>
Manu Sporny: +1<br/>
Pierre-Antoine Champin: +1<br/>
Rob Sanderson: +1<br/>
Benjamin Young: +1<br/>
Ivan Herman: +1<br/>
Gregg Kellogg: +1<br/>
Simon Steyskal: +1<br/>
Resolution #4: (2/) <code>@direction</code> is legal as part of a value object in the json-ld internal model but may not be used if <code>@type</code> is present*<br/>
Duy-Tung Luu: why no <code>@direction</code> with <code>@type?</code><br/>
Gregg Kellogg: because you can’t have <code>@language</code> with <code>@type</code><br/>
<b>tung:</b> but what if the <code>@type</code> has a language?<br/>
Gregg Kellogg: that’s a different case<br/>
<b>tung:</b> can we use a class to define language?<br/>
Gregg Kellogg: that was considered<br/>
*Proposed resolution: (3/) <code>@direction</code> is legal in an expanded term definition if you could have used <code>@language</code> in that definition and applies to all values of that term in the same way as <code>@language**</code> (Rob Sanderson)<br/>
Rob Sanderson: I think we’re back in solution space, and should get back to resolutions<br/>
Gregg Kellogg: +1<br/>
Manu Sporny: +1<br/>
Benjamin Young: +1<br/>
Pierre-Antoine Champin: +1<br/>
Ivan Herman: +1<br/>
<b>eg:</b> {"label": {"<code>@id":</code> "rdfs:label", "<code>@direction":</code> "rtl"}}<br/>
Resolution #5: (3/) <code>@direction</code> is legal in an expanded term definition if you could have used <code>@language</code> in that definition and applies to all values of that term in the same way as <code>@language*</code><br/>
*Proposed resolution: (4/) The value space of <code>@direction</code> is one of “ltr” and “rtl” (“auto” would have been the same as not asserting the keyword)** (Rob Sanderson)<br/>
Rob Sanderson: +1<br/>
Gregg Kellogg: +1<br/>
Ivan Herman: +1<br/>
Simon Steyskal: +1<br/>
Manu Sporny: +1<br/>
Benjamin Young: +1<br/>
Pierre-Antoine Champin: +1<br/>
Resolution #6: (4/) The value space of <code>@direction</code> is one of “ltr” and “rtl” (“auto” would have been the same as not asserting the keyword)*<br/>
*Proposed resolution: (5/) We expand the possible values of language maps to be either a string (as per today) or a value object with direction and without language or type** (Rob Sanderson)<br/>
Rob Sanderson: +1<br/>
Gregg Kellogg: +1<br/>
Pierre-Antoine Champin: +1<br/>
Ivan Herman: +1<br/>
Benjamin Young: +1<br/>
Manu Sporny: +1<br/>
<b>eg:</b> {“en”: {“<code>@value”:</code> “Rob”, “<code>@direction”:</code> “ltr”}}<br/>
Resolution #7: (5/) We expand the possible values of language maps to be either a string (as per today) or a value object with direction and without language or type*<br/>
*Proposed resolution: (6/) <code>@direction</code> can be present in a context node, and sets in the active context that the default direction for all string literals** (Rob Sanderson)<br/>
Rob Sanderson: +1<br/>
Gregg Kellogg: +1<br/>
Ivan Herman: +1<br/>
Manu Sporny: +1<br/>
Simon Steyskal: +1<br/>
Benjamin Young: +<br/>
Pierre-Antoine Champin: +1<br/>
Benjamin Young: +1<br/>
Resolution #8: (6/) <code>@direction</code> can be present in a context node, and sets in the active context that the default direction for all string literals*<br/>
Proposed resolution: (7 agrees that <code>@container:</code> <code>@direction</code> is not a useful feature for JSON-LD 1.1 (unlike <code>@container:</code> <code>@language)</code> (Rob Sanderson)<br/>
Rob Sanderson: +1<br/>
Manu Sporny: +1<br/>
Dan Brickley: +1<br/>
Pierre-Antoine Champin: +1<br/>
Benjamin Young: +1<br/>
Simon Steyskal: +1<br/>
Resolution #9: (7/) WG agrees that <code>@container:</code> <code>@direction</code> is not a useful feature for JSON-LD 1.1 (unlike <code>@container:</code> <code>@language)*</code><br/>
Ivan Herman: +1<br/>
</details>

**Comment by iherman** [comment]
This issue was discussed in [a meeting](https://www.w3.org/2018/json-ld-wg/Meetings/Minutes/2019/2019-09-20-json-ld#section1-2).

- `RESOLVED:  do not, by default, handle <code>@direction</code> expression in toRDF (or fromRDF)`
<details><summary><i class="differentiate">View the transcript</i></summary>
Ivan Herman: what we agreed yesterday is the syntax. We will have direction in json-ld. Within json-ld world it just works<br/>
… however, the question is if we generate RDF, what do we do?<br/>
Ralph Swick: can we be more precise about “generate RDF”?<br/>
Ivan Herman: Produce an RDF graph with a literal with a direction<br/>
… then one approach is that the processor has two api run-time options. One option is to say forget it. It would just disappear and generate a traditional language literal<br/>
… the other is to define new RDF terms for language and direction, where instead of a literal it’s a blank node<br/>
… a kind of skos-xl like structure for literals<br/>
… the positive aspect of the structure is that it round trips through the graph<br/>
… the information isn’t lost<br/>
… the negative aspect is if RDF data goes this way, we create a parallel universe for language+direction literals<br/>
… json-ld users can pick which way they want<br/>
… but what’s the default<br/>
… don’t have an answer, perhaps via a coin flip<br/>
… that’s where we are<br/>
… there are other possibilities for how to reflect into RDF<br/>
… could also be a new datatype<br/>
… there’s always a parallel universe<br/>
Rob Sanderson: [looks at rdf literals and base direction doc]<br/>
Ivan Herman: question is whether there’s a problem to define two new terms in rdf: namespace<br/>
Ralph Swick: are you already comfortable with the processor option?<br/>
Gregg Kellogg: there are options for other things, though most people use default<br/>
… Not acceptable to not round trip. Should do something that preserves it.<br/>
… lots of discussion. Future group would be expected to replace or endorse the structure. A processor option would let the future processor adapt<br/>
… can’t keep it entirely in the JSON-LD domain<br/>
Rob Sanderson: I generally agree. two comments, though<br/>
… for CR, if we mark it at risk, and it turns out that….<br/>
Ivan Herman: mark what?<br/>
Rob Sanderson: well…whatever we define here<br/>
… and at implementation phase it turns out there’s push back against the structured approach<br/>
… because it results in an “always punning” scenario<br/>
… and if, like us, you don’t have punning as an option<br/>
… then the language tag approach would be preferred<br/>
… which would be my personal preference<br/>
… because it puts things in an expected place<br/>
… the SPARQL queries already have to search for “en-*”<br/>
… there are of course other cons<br/>
… but if it turns out that during implementation that the structured approach is not the best default<br/>
… do we have to go back through CR in order to change that default?<br/>
Gregg Kellogg: the i18n seemed to strongly object to the use (well…“hacking”) of the language tag<br/>
Ivan Herman: I’d even expect them to formally object to that solution<br/>
… at this point I don’t think it’s a choice<br/>
Rob Sanderson: Addison did say he’d rather have it in the language string than to not keep it at all<br/>
Manu Sporny: I don’t think the structure solves things. We haven’t had this for a long time.<br/>
Ralph Swick: And getting objections as a result<br/>
Manu Sporny: Yes. We should be careful to do this right. It’s a half way solution that introduces problems. We should do the thing that we think is the final solution<br/>
Rob Sanderson: .. we can be a forcing function. By default it’s dropped, it doesn’t impact them if you don’t want it<br/>
Manu Sporny: so opt in<br/>
… what it looks like in RDF - I agree that the i18n folks dont’ want it in the tag - but there could be a ^direction<br/>
… of the options I’ve seen that looks the cleanest<br/>
… don’t have to upgrade infrastructure now until it’s implemented<br/>
… BCP47 doesn’t have a caret in any language<br/>
… so a clean token to split the string on<br/>
… if that’s the final option, then we should force it in that direction<br/>
Ivan Herman: Am fine with opt in<br/>
… but opt in should mean generate something in RDF which is valid today<br/>
… generating invalid RDF is unusable anyway<br/>
… no idea what the community will come to a solution, if ever, but something that no processor will accept today is useless<br/>
… caret means we create another syntax<br/>
… could have created something with a string and lots of carets bla bla :)<br/>
… won’t have it today, might not ever<br/>
… so generating it is pointless. Only one that’s valid is the structure<br/>
Pierre-Antoine Champin: If we go the way of an option, we should consider the equivalent for rdf to json-ld<br/>
… that might be none trivial for an intermediate node solution<br/>
Gregg Kellogg: to manu’s point, the algorithms to make a dataset rather than nquads serialization, adding a direction is something a dataset would need to store<br/>
… we have uris bnodes and literals<br/>
… only if language tagged can they have a language<br/>
… literal would have a direction and a language … you’ve added a column in a store<br/>
… an rdf WG might decide to do that<br/>
… deafening silence was the reaction from the community<br/>
Ralph Swick: strong +1 to opt in<br/>
… default should be to break the fewest tools<br/>
… option to do something better is most useful if the better is something that tools can work with<br/>
… suggests some choices<br/>
… don’t know how it works now but could have non binary option, and in 1.2 could then have a 3 or 4 value option with different syntaxes<br/>
… urge to do something useful to existing tools<br/>
Manu Sporny: What goes along with the option could be a function to decide what to output<br/>
… e.g. RDF in some format<br/>
… if you want new triples then great. BUt it lets the community to decide how to output it so they can integrate with different toolsets<br/>
… no clear solution that meets all the requirements<br/>
… if the solution is new triples, we have no interest in implementing it<br/>
… don’t have a strong requirement to do that, so would just wait until the problem has a final solution<br/>
… also to note that the argument of existing ecosystem not doing this is invoked a lot. Invoked in 1.0 but we defined various things and they were pulled into 1.1<br/>
… arguing this is not the state of the world today can be valid, but can’t push the ecosystem forward<br/>
Ralph Swick: [so if Manu has no strong requirement at the moment, having infinite extensibility with a function call-out seems way overblown]<br/>
Manu Sporny: concerned by implementation complexity. People saying they can’t implement a 1.1 processor<br/>
… every time they look at the spec they see a massive set of features<br/>
… looks like it will cost a lot of money so they pick and choose a small set and pay attention to the things they think are important<br/>
Proposed resolution: do not, by default, handle <code>@direction</code> expression in toRDF (or fromRDF) (Benjamin Young)<br/>
Manu Sporny: every new feature makes it worse – too many options and we make it very difficult.<br/>
Manu Sporny: +1<br/>
Rob Sanderson: Clarification - the default option for toRDF is to drop it<br/>
Benjamin Young: +1<br/>
Ivan Herman: +1<br/>
Manu Sporny: a future proposal will add in more options<br/>
Pierre-Antoine Champin: +0.5<br/>
Rob Sanderson: +1<br/>
Gregg Kellogg: +0.1<br/>
Resolution #1: do not, by default, handle <code>@direction</code> expression in toRDF (or fromRDF)<br/>
Rob Sanderson: from before – data type is also a valid option that solves the objectproperty / dataproperty distinction in OWL<br/>
</details>

**Comment by iherman** [comment]
This issue was discussed in [a meeting](https://www.w3.org/2018/json-ld-wg/Meetings/Minutes/2019/2019-09-20-json-ld#section4).

- `RESOLVED:  Add a parameter option to the API function toRDF, with three possible values: null or absent (do not process, the default), datatype (reflect to a datatype per 2.1.2.1.2) and bnode (reflect to a bnode structure per 2.1.4)`
- `RESOLVED:  Add a value to the options parameter to the API function fromRDF, with the above three options plus “any” which will convert via any conversion function available as discovered in the data`
<details><summary><i class="differentiate">View the transcript</i></summary>
Rob Sanderson: We agreed in the AM that we want a processor option, which defaults to drop the direction.<br/>
Ivan Herman: The datatype version familiy of language datatypes is also a viable option.<br/>
… 2.1.2.1.2 creates an ugly but valid datatype to describe language and direction #XX_YY.<br/>
… This wouldn’t be used in JSON-LD, but in the resulting quad.<br/>
… I now think this should be one option, and the other comound literal<br/>
Benjamin Young: This is adding switching in from/to RDF algorithms. 1.0 did not actually have API entrypoints for this.<br/>
… We could lower the bar for implementations by moving them into a separate interrace.<br/>
Pierre-Antoine Champin: I advocate the datatype solution over the compound node.<br/>
… The opt-in/out options would be different, but they would still be literals.<br/>
… In SPARQL, STR(literal) still gives you the expected results.<br/>
… Detecting these in the to/from algorithm would be easier and fewer corner-cases.<br/>
… On the semantics side, I think the DT solution makes it easy to define semantic interopration with existing language tags (intersecting value spaces).<br/>
… You could make interaction with existing lang-tagged strings smoother than as considering them as two different spieces of literals.<br/>
Ivan Herman: To be very precise, we’re talking about 2.1.2.1.2 in the document<br/>
Pierre-Antoine Champin: <a href="https://w3c.github.io/rdf-dir-literal/#datatype-family">https://w3c.github.io/rdf-dir-literal/#datatype-family</a><br/>
Rob Sanderson: I wanted to disuss 2.1.1.2.<br/>
… I think we could subclass rdf:langString to create rtl/ltr versions of that sub-type.<br/>
… Hm, Guess not, given RDF spec.<br/>
Pierre-Antoine Champin: I agree with gkellogg that the option in fromRdf might sound strange, but the message might not be clear. If there’s no option, we bake in the processing, and close the door for other possible options.<br/>
Benjamin Young: section references come from <a href="https://w3c.github.io/rdf-dir-literal/#compound-literal">https://w3c.github.io/rdf-dir-literal/#compound-literal</a><br/>
Proposed resolution: Add a parameter option to the API function toRDF, with three possible values: null or absent (do not process, the default), datatype (reflect to a datatype per 2.1.2.1.2) and bnode (reflect to a bnode structure per 2.1.4) (Rob Sanderson)<br/>
Rob Sanderson: +1<br/>
Benjamin Young: +1<br/>
Gregg Kellogg: +1<br/>
Ivan Herman: +1<br/>
Pierre-Antoine Champin: +1<br/>
Resolution #2: Add a parameter option to the API function toRDF, with three possible values: null or absent (do not process, the default), datatype (reflect to a datatype per 2.1.2.1.2) and bnode (reflect to a bnode structure per 2.1.4)<br/>
Benjamin Young: What if you have both bnodes and datatypes in an input? Could we convert them?<br/>
Proposed resolution: Add a value to the options parameter to the API function fromRDF, with the above three options plus “any” which will convert either datatype or bnode as discovered in the data (Rob Sanderson)<br/>
Rob Sanderson: +1<br/>
Ivan Herman: +1<br/>
Benjamin Young: +1<br/>
Gregg Kellogg: +1<br/>
Gregg Kellogg: For example `preserveDirection: enum(null, dt, combo, any)<br/>
Proposed resolution: Add a value to the options parameter to the API function fromRDF, with the above three options plus “any” which will convert via any conversion function available as discovered in the data (Rob Sanderson)<br/>
Pierre-Antoine Champin: I see this as an extension point, some may have more than two.<br/>
Ivan Herman: +1<br/>
Benjamin Young: +1<br/>
Gregg Kellogg: +1<br/>
Rob Sanderson: +1<br/>
Pierre-Antoine Champin: +1<br/>
Resolution #3: Add a value to the options parameter to the API function fromRDF, with the above three options plus “any” which will convert via any conversion function available as discovered in the data<br/>
</details>

**Comment by iherman** [comment]
This issue was discussed in [a meeting](https://www.w3.org/2018/json-ld-wg/Meetings/Minutes/2019/2019-10-04-json-ld#section4-1).

- `RESOLVED:  the WG is happy to merge #276, but leaves it to the editors whether they do it now or after the API changes is also in a PR`
- `ACTION:  create issue on values of language maps not including <code>@direction</code> (Gregg Kellogg)`
<details><summary><i class="differentiate">View the transcript</i></summary>
Benjamin Young: See Issue Syntax#1§1<br/>
Benjamin Young: See Pending PR on <code>@direction</code><br/>
Benjamin Young: no much to discuss; just to make sure everyone is aware of the PR<br/>
Gregg Kellogg: I will only merge it when the tests pass;<br/>
… currently working on my implementation;<br/>
… maybe only merge when we have an API PR.<br/>
… The syntax PR does not address the ability for a language map to contain value objects with <code>@value</code> and <code>@direction.</code><br/>
… Currently they only support strings.<br/>
… So there is no way to override the direction in a language map.<br/>
Ivan Herman: I think we can defer this to a future version.<br/>
Benjamin Young: is there an issue for this already?<br/>
Ivan Herman: raise an issue, so we can make it clear if we decide to postpone it.<br/>
Action #3: create issue on values of language maps not including <code>@direction</code> (Gregg Kellogg)<br/>
Dave Longley: the syntax doc proposes 2 or 3 ways to represent this in RDF in order to roundtrip.<br/>
… We should only recommend one.<br/>
… We would have to maintain them for many years.<br/>
… My preferred option would be to drop the <code>@direction</code> by default,<br/>
… with an option to serialize it in the way we think is the best for future RDF.<br/>
… That would force people to update their RDF lib,<br/>
… but that what we were expecting if a dedicated WG had been created.<br/>
Ivan Herman: half of your wishes are already fulfilled:<br/>
… by default, the <code>@direction</code> is dropped.<br/>
… We picked two options for roundtripping:<br/>
… * the i18n family of datatypes,<br/>
… * the “compound literal” approach (not a very good name).<br/>
… At the meeting in Fukuoka, even around the table,<br/>
… there was no clear consensus about which one was the best option.<br/>
… The idea is to let the community decide which one they want to use.<br/>
Dave Longley: Why not a 3rd option where we add a property to a literal?<br/>
Ivan Herman: because this is not valid RDF.<br/>
… The failure to create a WG shows that the RDF community is not willing to change the core of RDF.<br/>
… The advantage of the two proposed approaches is that they work with the deployed RDF infrastructure,<br/>
… even though they require some application-level knowledge to be used.<br/>
Dave Longley: What I’m hearing is “more technical depts, and no interop”<br/>
… If this is something that people need, and we introduce a hack to get it work,<br/>
… people will adopt the hack.<br/>
Ivan Herman: you can call it a hack, but this is the only thing that people seem ready to accept.<br/>
… Trying to force RDF to change is a bigger hack in my opinion.<br/>
Dave Longley: JSON-LD 1.0 did influence the RDF WG at the time, to extend RDF.<br/>
Ivan Herman: this is the big difference; there was an RDF WG at the time.<br/>
… There is none at the moment, nor can I foresee one in the near future.<br/>
Dave Longley: sometimes, standard follow implementations, rather than the opposite.<br/>
Pierre-Antoine Champin: I’d argue that the proposed solutions are not so hacky from and RDF point of view<br/>
… I prefer the datatype solution<br/>
… but none of them seem a scandal to me<br/>
… if there was an RDF WG working on this, they would consider one of these as the future standard<br/>
Ivan Herman: +1 to pchampin<br/>
Gregg Kellogg: I agree, these solutions are reasonable from the point of view of RDF.<br/>
… Also, the default option is to drop direction.<br/>
… The majority of data will not use direction, and not produce RDF differently than today.<br/>
Dave Longley: dropping it is a good default;<br/>
… but I expect Digital Bazaar being force to keep it in some way.<br/>
… For example, in VC, when the graph is signed, we need to keep the whole information.<br/>
Gregg Kellogg: from a digital signature point of view,<br/>
… I would not consider the direction to be relevant.<br/>
… Would two values differing only by direction be really different?<br/>
Dave Longley: this is scary; that could be turned into a terrible vulnerability.<br/>
Ivan Herman: I would like a resolution to allow Greg to merge the direction PR;<br/>
… this has lingered for too long.<br/>
Proposed resolution: the WG is happy to merge #276, but leaves it to the editors whether they do it now or after the API changes is also in a PR (Ivan Herman)<br/>
Ivan Herman: +1<br/>
Gregg Kellogg: +1<br/>
Harold Solbrig: +1<br/>
Pierre-Antoine Champin: +1<br/>
Benjamin Young: +1<br/>
Tim Cole: +1<br/>
Dave Longley: +0<br/>
Ruben Taelman: +1<br/>
Resolution #2: the WG is happy to merge #276, but leaves it to the editors whether they do it now or after the API changes is also in a PR<br/>
</details>

**Comment by iherman** [comment]
With base direction added to the spec the issue is now part of the official [I18N review](https://github.com/w3c/json-ld-wg/issues/93); closing this one.
